crates/red_knot_python_semantic/src/types/signatures.rs

```diff
@@ -105,42 +95,209 @@ impl<'db> Parameters<'db> {
         } = parameters;
         let positional_only = posonlyargs
             .iter()
-            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
-            .collect();
-        let positional_or_keyword = args
-            .iter()
-            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
-            .collect();
+            .map(|arg| Parameter::PositionalOnly(ParamWithDefault::from_node(db, definition, arg)));
+        let positional_or_keyword = args.iter().map(|arg| {
+            Parameter::PositionalOrKeyword(ParamWithDefault::from_node(db, definition, arg))
+        });
         let variadic = vararg
             .as_ref()
-            .map(|arg| Parameter::from_node(db, definition, arg));
+            .map(|arg| Parameter::Variadic(Param::from_node(db, definition, arg)));
         let keyword_only = kwonlyargs
             .iter()
-            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
-            .collect();
+            .map(|arg| Parameter::KeywordOnly(ParamWithDefault::from_node(db, definition, arg)));
         let keywords = kwarg
             .as_ref()
-            .map(|arg| Parameter::from_node(db, definition, arg));
-        Self {
-            positional_only,
-            positional_or_keyword,
-            variadic,
-            keyword_only,
-            keywords,
+            .map(|arg| Parameter::Keywords(Param::from_node(db, definition, arg)));
+        Self(
+            positional_only
+                .chain(positional_or_keyword)
+                .chain(variadic)
+                .chain(keyword_only)
+                .chain(keywords)
+                .collect(),
+        )
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub(crate) fn iter(&self) -> std::slice::Iter<Parameter<'db>> {
+        self.0.iter()
+    }
+
+    /// Iterate initial positional parameters, not including variadic parameter, if any.
+    ///
+    /// For a valid signature, this will be all positional parameters. In an invalid signature,
+    /// there could be non-initial positional parameters; effectively, we just won't consider those
+    /// to be positional, which is fine.
+    pub(crate) fn positional_only(&self) -> impl Iterator<Item = &Parameter<'db>> {
+        self.iter().take_while(|param| param.is_positional())
+    }
+
+    /// Return number of positional parameters this signature can accept.
+    ///
+    /// Doesn't account for variadic parameter.
+    pub(crate) fn positional_count(&self) -> usize {
+        self.positional_only().count()
+    }
```

Comment:
to me this seems simple enough that I would just inline it at callsites :P but no strong opinion

Reply:
Ok, sure. I also noticed looking at this that the name `positional_only` for the iterator-returning method is bad, due to confusion with positional-only parameters. This method iterates only positional parameters, which is not the same thing as returning only positional-only parameters 😆 Renamed it to just `positional`